transports/janus_websockets.c

```diff
@@ -924,17 +982,57 @@ static int janus_websockets_admin_callback(struct libwebsocket_context *this,
+					if(sent > -1 && sent < ws_client->bufpending) {
+						/* We still couldn't send everything that was left, we'll try and complete this in the next round */
+						ws_client->bufpending -= sent;
+						ws_client->bufoffset += sent;
+					} else {
+						/* Clear the pending/partial write queue */
+						ws_client->bufpending = 0;
+						ws_client->bufoffset = 0;
+					}
+					/* Done for this round, check the next response/notification later */
+					libwebsocket_callback_on_writable(this, wsi);
+					janus_mutex_unlock(&ws_client->mutex);
+					return 0;
+				}
+				/* Shoot all the pending messages */
 				char *response = g_async_queue_try_pop(ws_client->messages);
 				if(response && !ws_client->destroy && !g_atomic_int_get(&stopping)) {
 					/* Gotcha! */
-					unsigned char *buf = g_malloc0(LWS_SEND_BUFFER_PRE_PADDING + strlen(response) + LWS_SEND_BUFFER_POST_PADDING);
-					memcpy(buf+LWS_SEND_BUFFER_PRE_PADDING, response, strlen(response));
-					JANUS_LOG(LOG_VERB, "Sending WebSocket message (%zu bytes)...\n", strlen(response));
-					int sent = libwebsocket_write(wsi, buf+LWS_SEND_BUFFER_PRE_PADDING, strlen(response), LWS_WRITE_TEXT);
-					JANUS_LOG(LOG_VERB, "  -- Sent %d/%zu bytes\n", sent, strlen(response));
-					g_free(buf);
-					g_free(response);
+					int buflen = LWS_SEND_BUFFER_PRE_PADDING + strlen(response) + LWS_SEND_BUFFER_POST_PADDING;
+					if(ws_client->buffer == NULL) {
+						/* Let's allocate a shared buffer */
+						JANUS_LOG(LOG_VERB, "[AdminWSS-%p] Allocating %d bytes (response is %zu bytes)\n", wsi, buflen, strlen(response));
+						ws_client->buflen = buflen;
+						ws_client->buffer = g_malloc0(buflen);
+					} else if(buflen > ws_client->buflen) {
+						/* We need a larger shared buffer */
+						JANUS_LOG(LOG_VERB, "[AdminWSS-%p] Re-allocating to %d bytes (was %d, response is %zu bytes)\n", wsi, buflen, ws_client->buflen, strlen(response));
+						ws_client->buflen = buflen;
+						ws_client->buffer = g_realloc(ws_client->buffer, buflen);
+					}
+					memcpy(ws_client->buffer + LWS_SEND_BUFFER_PRE_PADDING, response, strlen(response));
+					JANUS_LOG(LOG_VERB, "[AdminWSS-%p] Sending WebSocket message (%zu bytes)...\n", wsi, strlen(response));
+					int sent = libwebsocket_write(wsi, ws_client->buffer + LWS_SEND_BUFFER_PRE_PADDING, strlen(response), LWS_WRITE_TEXT);
+					JANUS_LOG(LOG_VERB, "[AdminWSS-%p]   -- Sent %d/%zu bytes\n", wsi, sent, strlen(response));
+					if(sent > -1 && sent < (int)strlen(response)) {
+						/* We couldn't send everything in a single write, we'll complete this in the next round */
+						ws_client->bufpending = strlen(response) - sent;
+						ws_client->bufoffset = LWS_SEND_BUFFER_PRE_PADDING + sent;
+						JANUS_LOG(LOG_VERB, "[AdminWSS-%p]   -- Couldn't write all bytes (%d missing), setting offset %d\n",
+							wsi, ws_client->bufpending, ws_client->bufoffset);
+					} else {
+						/* We can get rid of the message */
+						g_free(response);
```

Comment:
since response was already copied into buffer, and you'll complete partial writes with buffer, I think you should free "response" message in all cases